x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -0,0 +1,229 @@
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.gateway.GatewayService.STATE_NOT_RECOVERED_BLOCK;
+import static org.elasticsearch.xpack.core.ClientHelper.SEARCHABLE_SNAPSHOTS_ORIGIN;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_INDEX_ID_SETTING;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_INDEX_NAME_SETTING;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_SNAPSHOT_ID_SETTING;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_SNAPSHOT_NAME_SETTING;
+
+/**
+ * A service that delete documents in the snapshot blob cache index when they are not required anymore.
+ *
+ * This service runs on the data node that contains the snapshot blob cache primary shard. It listens to cluster state updates to find
+ * searchable snapshot indices that are deleted and checks if the index snapshot is still used by other searchable snapshot indices. If the
+ * index snapshot is not used anymore then i triggers the deletion of corresponding cached blobs in the snapshot blob cache index using a
+ * delete-by-query.
+ */
+public class BlobStoreCacheMaintenanceService implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(BlobStoreCacheMaintenanceService.class);
+
+    private final ClusterService clusterService;
+    private final Client clientWithOrigin;
+    private final String systemIndexName;
+    private final ThreadPool threadPool;
+
+    public BlobStoreCacheMaintenanceService(ClusterService clusterService, ThreadPool threadPool, Client client, String systemIndexName) {
+        this.clientWithOrigin = new OriginSettingClient(Objects.requireNonNull(client), SEARCHABLE_SNAPSHOTS_ORIGIN);
+        this.systemIndexName = Objects.requireNonNull(systemIndexName);
+        this.clusterService = Objects.requireNonNull(clusterService);
+        this.threadPool = Objects.requireNonNull(threadPool);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        final ClusterState state = event.state();
+        if (state.getBlocks().hasGlobalBlock(STATE_NOT_RECOVERED_BLOCK)) {
+            return; // state not fully recovered
+        }
+        if (event.indicesDeleted() == null || event.indicesDeleted().isEmpty()) {
+            return; // no indices deleted in this cluster state update
+        }
+        final ShardRouting primary = systemIndexPrimaryShard(state);
+        if (primary == null || Objects.equals(state.nodes().getLocalNodeId(), primary.currentNodeId()) == false) {
+            return; // system index primary shard does not exist or is not assigned to this data node
+        }
+
+        final Set<MaintenanceTask> tasks = new HashSet<>();
+
+        for (Index deletedIndex : event.indicesDeleted()) {
```

Comment:
Could we add this to the task spawned on generic pool, seems like we would always spawn at least one task here. And then perhaps just spawn one task, responsible for cleaning this rather than potentially many tasks? Just to take more of the burden off the applier thread.